example/src/main/java/io/netty/example/stomp/websocket/StompWebSocketFrameEncoder.java

```diff
@@ -16,53 +16,64 @@
 import io.netty.handler.codec.stomp.StompFrame;
+import io.netty.handler.codec.stomp.StompHeaders;
 import io.netty.handler.codec.stomp.StompHeadersSubframe;
 import io.netty.handler.codec.stomp.StompSubframe;
 import io.netty.handler.codec.stomp.StompSubframeEncoder;
 
 import java.util.List;
 
-public class StompWebSocketFrameEncoder extends StompSubframeEncoder {
+public class StompWebSocketFrameEncoder extends StompSubframeEncoder
+        implements StompEncodedSubframeConverter<WebSocketFrame> {
 
     @Override
-    public void encode(ChannelHandlerContext ctx, StompSubframe msg, List<Object> out) throws Exception {
+    public void encode(ChannelHandlerContext ctx, StompSubframe msg, List<Object> out) {
         super.encode(ctx, msg, out);
+    }
+
+    @Override
+    protected StompEncodedSubframeConverter<?> converter() {
+        return this;
+    }
 
-        if (out.isEmpty()) {
-            return;
+    @Override
+    public WebSocketFrame convert(StompFrame original, ByteBuf encoded) {
+        if (isTextFrame(original)) {
+            return new TextWebSocketFrame(encoded);
         }
 
-        final WebSocketFrame webSocketFrame;
-        if (msg instanceof StompFrame) {
-            if (out.size() == 1) {
-                webSocketFrame = new TextWebSocketFrame(getFirst(out));
-            } else {
-                CompositeByteBuf content = ctx.alloc().compositeBuffer(out.size());
-                for (Object byteBuf : out) {
-                    content.addComponent(true, (ByteBuf) byteBuf);
-                }
-                webSocketFrame = new TextWebSocketFrame(content);
-            }
-        } else if (msg instanceof StompHeadersSubframe) {
-            webSocketFrame = new TextWebSocketFrame(false, 0, getFirst(out));
-        } else if (msg instanceof LastStompContentSubframe) {
-            webSocketFrame = new ContinuationWebSocketFrame(true, 0, getFirst(out));
-        } else {
-            webSocketFrame = new ContinuationWebSocketFrame(false, 0, getFirst(out));
+        return new BinaryWebSocketFrame(encoded);
+    }
+
+    @Override
+    public WebSocketFrame convert(StompHeadersSubframe original, ByteBuf encoded) {
+        if (isTextFrame(original)) {
+            return new TextWebSocketFrame(false, 0, encoded);
+        }
+
+        return new BinaryWebSocketFrame(false, 0, encoded);
+    }
+
+    @Override
+    public WebSocketFrame convert(StompContentSubframe original, ByteBuf encoded) {
+        if (original instanceof LastStompContentSubframe) {
+            return new ContinuationWebSocketFrame(true, 0, encoded);
         }
 
-        out.clear();
-        out.add(webSocketFrame);
+        return new ContinuationWebSocketFrame(false, 0, encoded);
     }
 
-    private static ByteBuf getFirst(List<Object> container) {
-        return (ByteBuf) container.get(0);
+    private static boolean isTextFrame(StompHeadersSubframe headersSubframe) {
```

Comment:
This methods may be override by some one.

Reply:
This is just example for STOMP over WS you can easily write yourself encoder